### PR TITLE
Issue/2155 Item 6: Control Header Item & Content API via OPA

### LINKED ISSUE #2155: End to end basic "Approver" functionality: Item 6 Control Headers from OPA
### This is the break-down task `item 6` for  #2140 

Make sure `<Header>` component access `OPA` and filter our header items that the current user is not allowed to access.

> Item 6 Estimate: 3~5
> Blocked by Item 1, 2 & 3

### PR DESCRIPTION
**Pull request by t83714**
### What this PR does

Fixes #2155 

This PR includes:
- a more generic OPA result parser: `OpaCompileResponseParser`
usage: 
```typescript
import OpaCompileResponseParser from "@magda/typescript-common/dist/OpaCompileResponseParser";
const parser = new OpaCompileResponseParser();
parser.parse(data); // --- data can be json string or js object
// --- evaluateRule 
const result = parser.evaluateRule(
  "data.partial.object.content.allowRead"
);
// --- evaluateRule as human readable string --- more for testing 
const result = parser.evaluateRuleAsHumanReadableString(
  "data.partial.object.content.allowRead"
);
```

- Simple SQL translator: `SimpleOpaSQLTranslator`
usage:
```typescript
import OpaCompileResponseParser from "@magda/typescript-common/dist/OpaCompileResponseParser";
import SimpleOpaSQLTranslator from "@magda/typescript-common/dist/SimpleOpaSQLTranslator";
const parser = new OpaCompileResponseParser();
parser.parse(response);

const result = parser.evaluateRule(
    "data.partial.object.content.allowRead"
);

const sqlValues:any[] = [];
const translator = new SimpleOpaSQLTranslator(["input.object.content"]);
const sqlWhereClause:string = translator.parse(result, sqlValues);
```
- `sqlWhereClause` includes where cluse you can use to secure your sql query
- Please note: you should concat `sqlWhereClause` with other existing whereClause with `AND`
- `sqlValues` is a `any[]`. Translator will push values into this array when inject values into whereClause (ie. you don't have to always supply an empty array.  )

### Two ways to secure your API or resource

This PR demonstrates two ways to secure your API or resource:
- Compile & Partially evaluation. You don't need this very often. Only read access to more than one items/resource. You will need `OpaCompileResponseParser` & `SimpleOpaSQLTranslator` (or write your own translator) for this option. `magda-content-api/src/Database.ts / getContentSummary()` provides one example.
- Directly query a pre-defined policy. This is the more straight-forward way as:
  - The API output all control by your policy. Can be a simple `true` or `false` or an array of items
  - You don't need any extra libs for any further parsing or translation. Just an HTTP request to your policy.
  -  `magda-content-api/src/Database.ts / getContentById()` provides one example.

### Relavent Test Case

You can find more examples of parser from test case: `magda-typescript-common/src/test/testOpaCompileResponseParser.spec.ts`

### Todo:

Just realised the policy for content items doesn't have to be such complex --- We can re-write it in a more concise way 

#### Commits after issue/2153

https://github.com/magda-io/magda/compare/issue/2154...issue/2155

### Checklist

-   [x] There are unit tests to verify my changes are correct
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: magda-typescript-common/src/typings.d.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+declare module "*.json" {
```
Comment (AlexGilleran) [inline magda-typescript-common/src/typings.d.ts]: I think there's an option in tsconfig that does this for you now?

File: deploy/helm/magda/charts/opa/policies/object/content/allowRead.rego [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    hasAnyDraftReadPermission != true
+    hasAnyPublishedReadPermission != true
+    input.object.content.id != "header/navigation/drafts"
+    input.object.content.id != "header/navigation/datasets"
+}
+
+allowRead {
+    isValidAllowReadUri == true
+    contentId := getContentIdFromOperationUri(input.operationUri)
+    ## It's a pattern
```
[the remark this answers is not in the document]
Comment (t83714) [reply deploy/helm/magda/charts/opa/policies/object/content/allowRead.rego]: Added more comments

File: magda-typescript-common/src/OpaCompileResponseParser.ts [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+    parser: OpaCompileResponseParser;
+}
+
+export class RegoRule {
+    public name: string;
+    public fullName: string;
+    public isDefault: boolean;
+    public value: RegoValue;
+    public expressions: RegoExp[];
+    // --- whether this rule is matched
```
[the remark this answers is not in the document]
Comment (t83714) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: Added more document~

File: magda-typescript-common/src/OpaCompileResponseParser.ts [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+    name: string;
+    fullName: string;
+    isDefault: boolean;
+    value: RegoValue;
+    expressions: RegoExp[];
+    isCompleteEvaluated?: boolean;
+    parser: OpaCompileResponseParser;
+}
+
+export class RegoRule {
```
[the remark this answers is not in the document]
Comment (t83714) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: Added more document~

File: magda-typescript-common/src/OpaCompileResponseParser.ts [hunk 5]
```diff
@@ -0,0 +1,764 @@
...
+                        throw new Error(
+                            `Invalid 3 terms rego expression, Unknown operator "${operator}": ${this.termsAsString()}`
+                        );
+                }
+                this.isCompleteEvaluated = true;
+                this.value = value;
+                return this;
+            }
+        }
+        // --- so far there is no 2 terms expression e.g. ! x
```
[the remark this answers is not in the document]
Comment (t83714) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: Throw an error for non 1 or 3 terms now~

File: magda-typescript-common/src/typings.d.ts (hunk 1 above)
Comment (t83714) [reply magda-typescript-common/src/typings.d.ts]: 👍 Removed the `typings.d.ts`

**Comment by t83714** [comment]
@AlexGilleran It's ready to have a look again~
Let me know any further changes are required~

File: magda-typescript-common/src/OpaCompileResponseParser.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexGilleran) [reply magda-typescript-common/src/OpaCompileResponseParser.ts]: Oh wow you really went all out! Awesome :)

**Pull request merged**